### PR TITLE
feat(dynamic_samping): Adjust sample rate by envelope header's sample_rate [INGEST-1395]

### PR DESCRIPTION
**Pull request by untitaker**
We have a new envelope header that tells us the configured sample rate of the client that initiated the trace. Adjust trace sampling by that incoming sample rate.

TODO:

- [x] adjust transaction sample rate as well

### CODE REVIEW

File: relay-sampling/src/lib.rs [hunk 1]
```diff
@@ -759,9 +766,24 @@ impl DynamicSamplingContext {
...
+            if adjusted_sample_rate.is_infinite() || adjusted_sample_rate.is_nan() {
+                // adjusted_sample_rate is infinite or NaN. This is a bug.
+                //
+                // - infinite should not happen because we clamped the number.
+                // - NaN can happen if the client_sample_rate is 0, which is bogus because the SDK
+                //   should've dropped the envelope. In that case let's pretend the sample rate was
+                //   not sent, because clearly the sampling decision across the trace is still 1.
+                //   The most likely explanation is that the SDK is reporting its own sample rate
+                //   setting instead of the one from the continued trace.
+                adjusted_sample_rate = 1.0;
```
Comment (jjbayer) [inline relay-sampling/src/lib.rs]: Do we want to log an error for this?

File: relay-sampling/src/lib.rs [hunk 2]
```diff
@@ -2052,8 +2092,33 @@ mod tests {
...
+          "trace_id": "00000000-0000-0000-0000-000000000000",
+          "public_key": "abd0f232775f45feab79864e580d160b",
+          "release": null,
+          "environment": null,
+          "transaction": null,
+          "sample_rate": "0.5",
+          "user_id": "hello"
+        }
+        "###);
+    }
```
Comment (jjbayer) [inline relay-sampling/src/lib.rs:2225]: Should we test this with some more values for `sample_rate`? E.g. "bogus", "10e4", ...

File: relay-sampling/src/lib.rs [hunk 3]
```diff
@@ -759,9 +766,24 @@ impl DynamicSamplingContext {
...
+                // adjusted_sample_rate is infinite or NaN. This is a bug.
+                //
+                // - infinite should not happen because we clamped the number.
+                // - NaN can happen if the client_sample_rate is 0, which is bogus because the SDK
+                //   should've dropped the envelope. In that case let's pretend the sample rate was
+                //   not sent, because clearly the sampling decision across the trace is still 1.
+                //   The most likely explanation is that the SDK is reporting its own sample rate
+                //   setting instead of the one from the continued trace.
+                adjusted_sample_rate = 1.0;
+            }
```
Comment (jjbayer) [inline relay-sampling/src/lib.rs]: Even though the formula is simple, it would be nice to wrap this in a function `adjust_sample_rate()` with unit tests for a few cases (e.g. `(0,1), (1,1), (0.1, 0.5), (0.5, 0.1)` etc.).

File: relay-sampling/src/lib.rs [hunk 4]
```diff
@@ -745,6 +749,9 @@ pub struct DynamicSamplingContext {
     /// Set on transaction start, or via `scope.transaction`.
     #[serde(default)]
     pub transaction: Option<String>,
+    /// The same rate with which this trace was sampled. This is a float between 0 and 1.
+    #[serde(default)]
+    pub sample_rate: Option<JsonStringifiedValue<f64>>,
```
Comment (jjbayer) [inline relay-sampling/src/lib.rs]: Do we need an error boundary here? Or is it OK to fail the entire `trace` header instead?

File: relay-server/src/utils/dynamic_sampling.rs [hunk 5]
```diff
@@ -94,14 +102,6 @@ pub fn get_sampling_key(envelope: &Envelope) -> Option<ProjectKey> {
...
 }
 
-pub fn sample_trace(
-    envelope: &Envelope,
-    project_state: &ProjectState,
-    processing_enabled: bool,
-) -> Result<(), RuleId> {
-    sample_transaction_internal(envelope, project_state, processing_enabled)
-}
-
```
Comment (jjbayer) [inline relay-server/src/utils/dynamic_sampling.rs:104]: Thank you for this!

File: relay-sampling/src/lib.rs (hunk 1 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs]: logged a warning

File: relay-sampling/src/lib.rs (hunk 2 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs:2225]: done

File: relay-sampling/src/lib.rs (hunk 3 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs]: done

File: relay-sampling/src/lib.rs (hunk 4 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs]: I don't really know right now which behavior is best. if the sample rate is malformed like that, we will disable traces sampling, which we also do on other invalid fields

Comment (jjbayer) [reply relay-sampling/src/lib.rs]: OK that sounds fine.

File: relay-sampling/src/lib.rs [hunk 6]
```diff
@@ -745,23 +749,52 @@ pub struct DynamicSamplingContext {
...
+            // - infinite should not happen because we clamped the number.
+            // - NaN can happen if the client_sample_rate is 0, which is bogus because the SDK
+            //   should've dropped the envelope. In that case let's pretend the sample rate was
+            //   not sent, because clearly the sampling decision across the trace is still 1.
+            //   The most likely explanation is that the SDK is reporting its own sample rate
+            //   setting instead of the one from the continued trace.
+            //
+            // since we write back the client_sample_rate into the event's trace context, it should
+            // be possible to find those values + sdk versions via snuba
+            relay_log::warn!("adjusted sample rate ended up being inf/nan");
```
Comment (jjbayer) [inline relay-sampling/src/lib.rs]: nit: Not sure if we already have enough context on the sentry scope to identify the project or sdk version sending this, but we can also add that when the warning shows up (if it ever will).

**Review by jjbayer — APPROVED** [review]

File: relay-sampling/src/lib.rs (hunk 6 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs]: we don't have enough context here right now, i think it's easy enough to look into snuba

File: relay-common/src/serde.rs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+/// struct TraceContext {
+///     sample_rate: JsonStringifiedValue<f64>,
+/// }
+/// ```
+///
+/// ...deserializes from `{"sample_rate": "1.0"}` and `{"sample_rate": 1.0}`, and serializes to
+/// `{"sample_rate": "1.0"}`
+///
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+pub struct JsonStringifiedValue<T>(pub T);
```
Comment (jan-auer) [inline relay-common/src/serde.rs]: This type is rather complex for the single use, and allows direct deserialization which we do not actually want in the sampling context. Instead, it would be great if we just use a custom deserialize function or even parse this lazily in an accessor function.

File: relay-sampling/src/lib.rs (hunk 4 above)
Comment (jan-auer) [reply relay-sampling/src/lib.rs]: Let's fail attributes individually, for the same reasoning we do this in event payloads.

File: relay-sampling/src/lib.rs [hunk 8]
```diff
@@ -745,23 +749,52 @@ pub struct DynamicSamplingContext {
     /// Set on transaction start, or via `scope.transaction`.
     #[serde(default)]
     pub transaction: Option<String>,
+    /// The same rate with which this trace was sampled. This is a float between 0 and 1.
```
Comment (jan-auer) [inline relay-sampling/src/lib.rs]: Please add a `#[serde(flatten)] other: BTreeMap<String, String>` at the bottom to catch and forward unknown fields.

File: relay-general/Cargo.toml [hunk 9]
```diff
@@ -31,7 +31,7 @@ relay-general-derive = { path = "derive" }
 schemars = { version = "0.8.1", features = ["uuid", "chrono"], optional = true }
 sentry-release-parser = { version = "1.3.1" }
 serde = { version = "1.0.114", features = ["derive"] }
-serde_json = "1.0.55"
+serde_json = { version = "1.0.55", features = ["raw_value"] }
```
Comment (jan-auer) [inline relay-general/Cargo.toml]: This is no longer needed.

File: relay-server/src/actors/envelopes.rs [hunk 10]
```diff
@@ -1522,6 +1522,18 @@ impl EnvelopeProcessor {
             event._metrics = Annotated::new(metrics);
         }
 
+        if let Some(sampling_context) = envelope.sampling_context() {
+            if let Some(client_sample_rate) = sampling_context.sample_rate {
+                if let Some(ref mut contexts) = event.contexts.value_mut() {
+                    if let Some(EventContext::Trace(ref mut trace_context)) =
+                        contexts.get_context_mut("trace")
+                    {
+                        trace_context.client_sample_rate = Annotated::from(client_sample_rate.0);
```
Comment (jan-auer) [inline relay-server/src/actors/envelopes.rs]: Two things about this:

1. During sampling, let's ensure we update the envelope header with the new sample rate. This allows a Relay to detect if the previous pipeline (regardless of whether that was a Relay or SDK) already applied the correct sample rate. Effectively, this makes sampling idempotent. 
2. This here is merely for informational purposes, so it can & should be moved into store normalization.

If we need the actual _client_ sample rate in the event, this means we need two headers for this.

**Comment by untitaker** [comment]
@jan-auer okay, I addressed half of the review comments. I am not acting on two things in this PR:

* Sampling readjustment in case of relay chains. I looked into implementing this as we discussed offline and I now believe this actually is not possible at the moment without a larger refactor that merges trace-sampling and event-sampling into one function. That's a refactor I want to do anyway. But after this PR is merged.
* Failing attributes individually. If we want to do this we should do it for all attributes or have a real policy about this. Right now we don't, and throw the entire sampling context away. And IMO that's not even that wrong, after all we should probably not apply dynamic sampling at all if we can't parse basic attributes. We could however store the raw value in the ErrorBoundary and use it to serialize the value again.


I think the PR as-is gets us quite far already. It doesn't make the existing problem of double-sampling worse, and we can run demos in prod as long as we point SDKs to the apex domain.

File: relay-common/src/serde.rs [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    /// Deserialize a value from a nested json string
+    pub fn deserialize<'de, T, D>(deserializer: D) -> Result<T, D::Error>
+    where
+        for<'de2> T: Deserialize<'de2>,
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(untagged)]
+        enum Helper<T> {
+            Verbatim(T),
```
Comment (jan-auer) [inline relay-common/src/serde.rs]: Please remove the verbatim case, let's be strict here and just support the string. 

Also, let's move this directly into `relay-sampling` until we have a need to share this with other crates. There is a _similar_ function in Profiling already (see [here](https://github.com/getsentry/relay/blob/e5c80898363871a256073c511bc9ad8b17385e13/relay-server/src/utils/profile.rs#L37)), however it uses `parse()` rather than `serde_json` and it allows both deserializations, so you can't share the implementation.

Let's please keep this as simple as possible.

File: relay-sampling/src/lib.rs [hunk 12]
```diff
@@ -745,23 +749,56 @@ pub struct DynamicSamplingContext {
...
     pub user: TraceUserContext,
+    /// Spillover values for backwards/forwards compat
+    #[serde(flatten, default)]
+    pub other: BTreeMap<String, Value>,
 }
 
 impl DynamicSamplingContext {
+    /// Compute the effective sampling rate based on the random "diceroll" and the sample rate from
+    /// the matching rule.
+    pub fn adjust_sample_rate(&self, rule_sample_rate: f64) -> f64 {
```
Comment (jan-auer) [inline relay-sampling/src/lib.rs]: _nit_: This method does not adjust a sample rate, it actually gives you a compensated sample rate back. The active "adjust" suggests that something is mutated. For instance, you could go with `adjusted_sample_rate` or `remaining_sample_rate` or something like that.

**Review by jan-auer — COMMENTED** [review]
Requires the clippy fix in #1329.

File: relay-general/src/protocol/contexts.rs [hunk 13]
```diff
@@ -496,6 +496,12 @@ pub struct TraceContext {
     /// excluding its immediate child spans.
     pub exclusive_time: Annotated<f64>,
 
+    /// The SDK-side sample rate as reported in the envelope's `trace.sample_rate` header.
+    ///
+    /// The server takes this field from envelope headers and writes it back into the event. SDKs
+    /// should not ever send this value.
```
Comment (iker-barriocanal) [inline relay-general/src/protocol/contexts.rs]: nit:
```suggestion
    /// The client-side sample rate as reported in the envelope's `trace.sample_rate` header.
    ///
    /// The server takes this field from envelope headers and writes it back into the event. Clients
    /// should not ever send this value.
```

File: relay-server/src/actors/envelopes.rs [hunk 14]
```diff
@@ -1522,6 +1522,18 @@ impl EnvelopeProcessor {
             event._metrics = Annotated::new(metrics);
         }
 
+        if let Some(sampling_context) = envelope.sampling_context() {
+            if let Some(client_sample_rate) = sampling_context.sample_rate {
+                if let Some(ref mut contexts) = event.contexts.value_mut() {
+                    if let Some(EventContext::Trace(ref mut trace_context)) =
+                        contexts.get_context_mut("trace")
+                    {
+                        trace_context.client_sample_rate = Annotated::from(client_sample_rate);
```
Comment (iker-barriocanal) [inline relay-server/src/actors/envelopes.rs]: The docs for `client_sample_rate` explicitly say SDKs shouldn't send this, but I wonder what relay's behavior should be if that happens. Has another relay already extracted this? Do we want to always smash the value there? Do we trust that other value, and ignore the value we have in the envelope?

This may have already been discussed and/or written somewhere. If this is the case, I'd add a brief comment in code that we are ok with smashing the value we get.

File: relay-sampling/src/lib.rs [hunk 15]
```diff
@@ -745,23 +798,60 @@ pub struct DynamicSamplingContext {
     /// Set on transaction start, or via `scope.transaction`.
     #[serde(default)]
     pub transaction: Option<String>,
+    /// The sample rate with which this trace was sampled in the SDK. This is a float between 0 and
+    /// 1.
```
Comment (iker-barriocanal) [inline relay-sampling/src/lib.rs]: nit:
```suggestion
    /// The sample rate with which this trace was sampled in the client. This is a float between 0 and
    /// 1.
```

File: relay-sampling/src/lib.rs [hunk 16]
```diff
@@ -745,23 +798,60 @@ pub struct DynamicSamplingContext {
...
+    /// Compute the effective sampling rate based on the random "diceroll" and the sample rate from
+    /// the matching rule.
+    pub fn adjusted_sample_rate(&self, rule_sample_rate: f64) -> f64 {
+        let client_sample_rate = self.sample_rate.unwrap_or(1.0);
+        let mut adjusted_sample_rate = (rule_sample_rate / client_sample_rate).clamp(0.0, 1.0);
+
+        if adjusted_sample_rate.is_infinite() || adjusted_sample_rate.is_nan() {
+            // adjusted_sample_rate is infinite or NaN. This is a bug.
+            //
+            // - infinite should not happen because we clamped the number.
```
Comment (iker-barriocanal) [inline relay-sampling/src/lib.rs]: When does `infinite` happen? I may be using a different version in the following playground, but the division panics: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c7315968753f0f27ead5b1fed0b0d22d

File: relay-sampling/src/lib.rs (hunk 6 above)
Comment (iker-barriocanal) [reply relay-sampling/src/lib.rs]: I'd handle each case separately, I think it could ease debugging with the appropriate SDK team.

File: relay-sampling/src/lib.rs [hunk 17]
```diff
@@ -745,23 +798,60 @@ pub struct DynamicSamplingContext {
...
+            // - NaN can happen if the client_sample_rate is 0, which is bogus because the SDK
+            //   should've dropped the envelope. In that case let's pretend the sample rate was
+            //   not sent, because clearly the sampling decision across the trace is still 1.
+            //   The most likely explanation is that the SDK is reporting its own sample rate
+            //   setting instead of the one from the continued trace.
```
Comment (iker-barriocanal) [inline relay-sampling/src/lib.rs]: I think handling each case separately could simplify the code, avoid panics in division by 0s (see comment above), and make it easier to identify and route the impact on Sentry:
```
if client_sample_rate == 0 {
  relay_log::warn(...);  // This is a client bug, let's not mix it with a weird case
  return 1;
} else {
  let adjusted = (rule_rate / client_rate).clamp(...);
  if adjusted.is_infinite() {
    // This should never happen
    relay_log::error(...);  // This should never happen, so the severity is higher than in client bugs
    return 1;
  }
  return adjusted;
}
```

File: relay-sampling/src/lib.rs (hunk 2 above)
Comment (iker-barriocanal) [reply relay-sampling/src/lib.rs:2225]: Can we also add a test for a negative `sample_rate`?

**Review by iker-barriocanal — DISMISSED** [review]
Blocking for panicking division by 0.

File: relay-sampling/src/lib.rs (hunk 16 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs]: it panicks with integers and returns infinite with floats

Comment (iker-barriocanal) [reply relay-sampling/src/lib.rs]: I see hehe, unblocking this then 👍

**Review by iker-barriocanal — COMMENTED** [review]
Removing my blocking, division by 0 doesn't panic with floats.

File: relay-server/src/actors/envelopes.rs (hunk 14 above)
Comment (untitaker) [reply relay-server/src/actors/envelopes.rs]: I decided this on my own, I don't think we have enough information to really understand what behavior is best. I however noticed just now that the code comment said that we should ignore the value from the client, but what my PR does is to force the effective sample rate to 1.0. Fixing that.

File: relay-sampling/src/lib.rs (hunk 17 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs]: Yeah I think that's better, fixing

File: relay-common/Cargo.toml [hunk 18]
```diff
@@ -23,6 +23,7 @@ relay-log = { path = "../relay-log" }
 sentry-types = "0.20.0"
 schemars = { version = "0.8.1", features = ["uuid", "chrono"], optional = true }
 serde = { version = "1.0.114", features = ["derive"] }
+serde_json = "1.0.55"
```
Comment (jan-auer) [inline relay-common/Cargo.toml]: This can be reverted now that there's no more code dependency on serde_json.

File: relay-general/src/store/mod.rs [hunk 19]
```diff
@@ -61,6 +61,9 @@ pub struct StoreConfig {
 
     /// Emit additional span attributes based on given configuration.
     pub span_attributes: BTreeSet<SpanAttribute>,
+
+    /// The SDK's sample rate as communicated via envelope headers
```
Comment (jan-auer) [inline relay-general/src/store/mod.rs]:
```suggestion
    /// The SDK's sample rate as communicated via envelope headers.
```

File: relay-general/src/store/normalize.rs [hunk 20]
```diff
@@ -134,6 +134,18 @@ impl<'a> NormalizeProcessor<'a> {
         }
     }
 
+    fn normalize_trace_context(&self, event: &mut Event) {
+        if let Some(client_sample_rate) = self.config.client_sample_rate {
```
Comment (jan-auer) [inline relay-general/src/store/normalize.rs]: _nit_: Rather than `if let`, you could always overwrite the value that's already in there, defaulting to `None`. That would enforce the constraint of the doc comment on `client_sample_rate` mentioning that clients are not allowed to send this value.

File: relay-sampling/src/lib.rs [hunk 21]
```diff
@@ -726,7 +726,60 @@ impl<'de> Deserialize<'de> for TraceUserContext {
...
 
+mod sample_rate_as_string {
+    use super::*;
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<f64>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // be super-strict with what kind of integers are accepted
+        let value = match Option::<String>::deserialize(deserializer)? {
```
Comment (jan-auer) [inline relay-sampling/src/lib.rs]: _nit_: If you want to squeeze the last bit of performance, deserialize into a `Cow` here.

File: relay-sampling/src/lib.rs [hunk 22]
```diff
@@ -745,23 +798,63 @@ pub struct DynamicSamplingContext {
...
+        default,
+        with = "sample_rate_as_string",
+        skip_serializing_if = "Option::is_none"
+    )]
+    pub sample_rate: Option<f64>,
     /// The user specific identifier ( e.g. a user segment, or similar created by the SDK
     /// from the user object).
     #[serde(flatten, default)]
     pub user: TraceUserContext,
+    /// Spillover values for backwards/forwards compat
```
Comment (jan-auer) [inline relay-sampling/src/lib.rs:814]: Our usual doc comment for this is:

```suggestion
    /// Additional arbitrary fields for forwards compatibility.
```

File: relay-sampling/src/lib.rs [hunk 23]
```diff
@@ -745,23 +798,63 @@ pub struct DynamicSamplingContext {
...
+            // that the SDK is reporting its own sample rate setting instead of the one from the
+            // continued trace.
+            //
+            // since we write back the client_sample_rate into the event's trace context, it should
+            // be possible to find those values + sdk versions via snuba
+            relay_log::warn!("client sample rate is <= 0");
+            rule_sample_rate
+        } else {
+            let adjusted_sample_rate = (rule_sample_rate / client_sample_rate).clamp(0.0, 1.0);
+            if adjusted_sample_rate.is_infinite() || adjusted_sample_rate.is_nan() {
```
Comment (jan-auer) [inline relay-sampling/src/lib.rs:838]: _nit_: The check for infinite is redundant due to `clamp` before.

**Review by jan-auer — APPROVED** [review]

File: relay-sampling/src/lib.rs (hunk 21 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs]: in a previous version, this actually makes it fail deserialization and I haven't figured out why. i will try it again

File: relay-sampling/src/lib.rs (hunk 23 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs:838]: it can be either nan if the `rule_sample_rate` is that, in either case it would be a bug so I'd rather check both in prod

File: relay-sampling/src/lib.rs (hunk 22 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs:814]: done

File: relay-sampling/src/lib.rs (hunk 21 above)
Comment (untitaker) [reply relay-sampling/src/lib.rs]: seems to work!

File: relay-general/src/store/normalize.rs (hunk 20 above)
Comment (untitaker) [reply relay-general/src/store/normalize.rs]: done
